src/edgehog_ota.c

```diff
@@ -42,331 +42,479 @@
     }
 
-    // Beware this function blocks the caller until OTA is completed.
-    return do_ota(edgehog_device, request_uuid, ota_url);
+    // Step 2 Perform the requested Update or Cancel operation. For the Update operation spawn
+    // a new task.
+
+    if (strcmp("Update", ota_operation) == 0) {
+        if (xTaskGetHandle("OTA UPDATE TASK") != NULL) {
+            pub_ota_event(
+                edgehog_dev, req_uuid, OTA_EVENT_FAILURE, 0, EDGEHOG_ERR_OTA_ALREADY_IN_PROGRESS);
+            return EDGEHOG_ERR_OTA_ALREADY_IN_PROGRESS;
+        }
+        ota_task_data.edgehog_dev = edgehog_dev;
+        ota_task_data.req_uuid = req_uuid;
+        ota_task_data.ota_url = ota_url;
+        TaskHandle_t ota_task_handle = NULL;
+        BaseType_t ota_task_ret = xTaskCreate(ota_task_code, "OTA UPDATE TASK", 4096,
+            &ota_task_data, tskIDLE_PRIORITY, &ota_task_handle);
+        if (ota_task_ret != pdPASS) {
+            ESP_LOGE(TAG, "OTA update task creation failed.");
+            pub_ota_event(edgehog_dev, req_uuid, OTA_EVENT_FAILURE, 0, EDGEHOG_ERR_OTA_INTERNAL);
+            esp_event_post(EDGEHOG_EVENTS, EDGEHOG_OTA_FAILED_EVENT, NULL, 0, 0);
+            return EDGEHOG_ERR_TASK_CREATE;
+        }
+    } else if (strcmp("Cancel", ota_operation) == 0) {
+        TaskHandle_t ota_task = xTaskGetHandle("OTA UPDATE TASK");
+        if (ota_task == NULL) {
+            pub_ota_event(
+                edgehog_dev, req_uuid, OTA_EVENT_FAILURE, 0, EDGEHOG_ERR_OTA_INVALID_REQUEST);
+            return EDGEHOG_ERR_OTA_INVALID_REQUEST;
+        }
+        xTaskNotify(ota_task, 0, eNoAction);
+    } else {
+        pub_ota_event(edgehog_dev, req_uuid, OTA_EVENT_FAILURE, 0, EDGEHOG_ERR_OTA_INVALID_REQUEST);
+        return EDGEHOG_ERR_OTA_INVALID_REQUEST;
+    }
+    return EDGEHOG_OK;
 }
 
-// TODO: make this function async using advanced_esp_ota instead of esp_https_ota
-static edgehog_err_t do_ota(
-    edgehog_device_handle_t edgehog_device, const char *request_uuid, const char *ota_url)
+/************************************************
+ *         Static functions definitions         *
+ ***********************************************/
+
+static void ota_task_code(void *pvParameters)
 {
-    ESP_LOGI(TAG, "INIT");
+    ota_task_data_t *task_data = (ota_task_data_t *) pvParameters;
+    edgehog_device_handle_t edgehog_dev = ((ota_task_data_t *) pvParameters)->edgehog_dev;
+    const char *req_uuid = ((ota_task_data_t *) pvParameters)->req_uuid;
+    nvs_handle_t handle_nvs;
+
+    // Step 1 acknowledge the valid update request and notify the start of the download operation.
+
+    pub_ota_event(edgehog_dev, req_uuid, OTA_EVENT_ACKNOWLEDGED, 0, EDGEHOG_OK);
+    pub_ota_event(edgehog_dev, req_uuid, OTA_EVENT_DOWNLOADING, 0, EDGEHOG_OK);
+
+    // Step 2 Open NVS namespace for the OTA update
+
+    ESP_LOGI(TAG, "OTA INIT");
     esp_event_post(EDGEHOG_EVENTS, EDGEHOG_OTA_INIT_EVENT, NULL, 0, 0);
```

Comment:
IMHO it is better to move `downloading` publish below `esp_event_post(EDGEHOG_OTA_INIT_EVENT)` the ACKNOWLEDGED status is already an init phase.
```suggestion

    // Step 2 Open NVS namespace for the OTA update

ESP_LOGI(TAG, "OTA INIT");
esp_event_post(EDGEHOG_EVENTS, EDGEHOG_OTA_INIT_EVENT, NULL, 0, 0);
pub_ota_event(edgehog_dev, req_uuid, OTA_EVENT_DOWNLOADING, 0, EDGEHOG_OK);
```